router/http/src/main/scala/io/buoyant/router/http/TracingFilter.scala

```diff
@@ -1,43 +1,83 @@
 package io.buoyant.router.http
 
-import com.twitter.finagle.{Service, SimpleFilter}
-import com.twitter.finagle.http.{Request, Response}
+import com.twitter.conversions.storage._
+import com.twitter.finagle.{Service, ServiceFactory, SimpleFilter, Stack, Stackable, param}
+import com.twitter.finagle.client.StackClient
+import com.twitter.finagle.http.{Request, Response, Status}
 import com.twitter.finagle.tracing.Trace
 
+object TracingFilter {
+  val role = StackClient.Role.protoTracing
+  val module: Stackable[ServiceFactory[Request, Response]] =
+    new Stack.Module1[param.Tracer, ServiceFactory[Request, Response]] {
+      val role = TracingFilter.role
+      val description = "Traces HTTP-specific request metadata"
+      def make(_tracer: param.Tracer, next: ServiceFactory[Request, Response]) = {
+        val param.Tracer(tracer) = _tracer
+        if (tracer.isNull) next
+        else (new TracingFilter).andThen(next)
+      }
+    }
+
+  val TransferEncoding = "transfer-encoding"
+}
+
 /**
  * Annotates HTTP method, uri, and status code, content-type, and content-length.
  */
-object TracingFilter extends SimpleFilter[Request, Response] {
+class TracingFilter extends SimpleFilter[Request, Response] {
+  import TracingFilter.TransferEncoding
 
   def apply(req: Request, service: Service[Request, Response]) = {
     recordRequest(req)
     service(req).onSuccess(recordResponse)
   }
 
-  private[this] def recordRequest(req: Request): Unit = {
+  private[this] def recordRequest(req: Request): Unit = if (Trace.isActivelyTracing) {
     Trace.recordRpc(req.method.toString)
     // http.uri is used here for consistency with finagle-http's tracing filter
     Trace.recordBinary("http.uri", req.uri)
     Trace.recordBinary("http.req.method", req.method.toString)
-    req.host.foreach(Trace.recordBinary("http.req.host", _))
+    req.host match {
+      case Some(h) => Trace.recordBinary("http.req.host", h)
+      case None =>
+    }
```

Comment:
We don't have to eliminate them entirely, but as in finagle, we should _avoid_ using anonymous functions in the request path. We know it creates more objects, and we should strive to minimize allocation.  Don't fear imperative code ;)  They're fine to use in other places (i.e. configuration code, admin interfaces, etc), but we need to be disciplined about allocation.